In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2200.170
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed bhi
bogomips	: 4400.34
clflush size	: 64
cache_alignment	: 64
ad

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

1
2
3

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
9

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py


^C


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

^C


### Test 1: Small numbers

In [ ]:
%%writefile input0.txt
10
0 0 0 0 0 0 0 0 0 0

Writing input0.txt


In [ ]:
%%writefile output0.txt
0

Writing output0.txt


In [ ]:
!./max_pairwise_product_full< input0.txt > ctest0.txt
!grep -xvFf output0.txt ctest0.txt

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product < input1.txt

/bin/bash: line 1: ./max_pairwise_product: No such file or directory


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [ ]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


### Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [ ]:
%%writefile input3.txt
6
1 3000 2000 9 20 23

Overwriting input3.txt


In [ ]:
%%writefile input4.txt
7
11 456 789 101112 131415 161718 192021

Writing input4.txt


In [ ]:
%%writefile input5.txt
5
951 753 654 852 379

Writing input5.txt


In [ ]:
%%writefile input9.txt
6
1 3333 2000 9 2222 23

Writing input9.txt


In [ ]:
%%writefile input10.txt
6
1111 2222 3333 4444 5555 6666

Writing input10.txt


In [ ]:
!./max_pairwise_product_long_full < input1.txt
!./max_pairwise_product_long_full < input2.txt
!./max_pairwise_product_long_full < input3.txt > output3.txt
!./max_pairwise_product_long_full < input4.txt > output4.txt
!./max_pairwise_product_long_full < input5.txt > output5.txt
!./max_pairwise_product_long_full < input6.txt > output6.txt
!./max_pairwise_product_long_full < input7.txt > output7.txt
!./max_pairwise_product_long_full < input8.txt > output8.txt
!./max_pairwise_product_long_full < input9.txt > output9.txt
!./max_pairwise_product_long_full < input10.txt > output10.txt

90
90
/bin/bash: line 1: input6.txt: No such file or directory
/bin/bash: line 1: input7.txt: No such file or directory
/bin/bash: line 1: input8.txt: No such file or directory


### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Writing output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000  real	4m56.082s user	4m53.554s sys	0m0.086s
^C


In [77]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

2147483646

real	5m8.024s
user	5m4.727s
sys	0m0.083s


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

570


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [ ]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [ ]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

571


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

200003177092
^C


In [ ]:
# Estimated time by instruction
# use the real reported  !time ./max_pairwise_product_long_full< input11.txt in seconds intead of the 295

295/200003177092


1.4749765693187072e-09

In [ ]:
# Estimated number of instructions by second

200003177092/295

677976871.4983051

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
10999 * 11000

120989000

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.856s
user	0m0.851s
sys	0m0.001s


## Python

In [78]:
%%writefile max_pairwise_product_full.py
n = int(input())                        #1
list = input().split()                  #1
list = [int(x) for x in list]           #n

maxProduct = 0                          #1
auxProduct = 0                          #1
for i in range(0,n):                    #1+2n
  for j in range(0,n):                  #n + 2n**2
    if i!=j:                            #n**2
      auxProduct = list[i]*list[j]      #n**2 -n
    if maxProduct < auxProduct:         #n**2 -n
      maxProduct = auxProduct           #[1, n**2/2 - n/2] upper bound

                                        # --------------
print(maxProduct)                       #5 + 3n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound

Writing max_pairwise_product_full.py


In [79]:
!chmod +x max_pairwise_product_full_py

!python max_pairwise_product_full.py < input3.txt > output3.txt
!python max_pairwise_product_full.py < input4.txt > output4.txt
!python max_pairwise_product_full.py < input5.txt > output5.txt
!python max_pairwise_product_full.py < input6.txt > output6.txt
!python max_pairwise_product_full.py < input7.txt > output7.txt
!python max_pairwise_product_full.py < input8.txt > output8.txt
!python max_pairwise_product_full.py < input9.txt > output9.txt
!python max_pairwise_product_full.py < input10.txt > output10.txt

chmod: cannot access 'max_pairwise_product_full_py': No such file or directory
/bin/bash: line 1: input6.txt: No such file or directory
/bin/bash: line 1: input7.txt: No such file or directory
/bin/bash: line 1: input8.txt: No such file or directory


In [ ]:
# Verify

# For the follwing solutions rquired  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [80]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [81]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


The number of instructions executed by the algorithm with a list of size $n$ is

$$T=11 + 7n + [2(n-1)n+1, 2.5 (n-1)n)$$


It depends on the size of the lists and the order of the list.

For the input1.txt

10


1 2 3 4 5 6 7 8 9 10

$$T(10)\in [109,306]$$


In [82]:
n = 10
print(11 + 7*10+2*(n-1)+n,11 + 7*10+2.5*(n-1)*n)

109 306.0


In [83]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt

90


**Python**

In [84]:
%%writefile max_pairwise_product.py
n = int(input())                        #1
list = input().split()                  #1
list = [int(x) for x in list]           #n

maxProduct = 0                          #1
auxProduct = 0                          #1
for i in range(0,n):                    #1+2n
  for j in range(i+1,n):                #((n**2)-n)/2
    if i!=j:                            #((n**2)-n)/2
      auxProduct = list[i]*list[j]      #((n**2)-n)/2
    if maxProduct < auxProduct:         #((n**2)-n)/2
      maxProduct = auxProduct           #[0,((n**2)-n)/2] upper bound

                                        # --------------
print(maxProduct)                       #5 + n + [5*n**2+1, 11*n**2/2  - n/2]

Writing max_pairwise_product.py


Creating big input

In [85]:
fileName = "inputPy.txt"

with open(fileName, "w") as file:
    file.write("20000"+ "\n")
    for i in range(1,20001):
        file.write(str(i) + " ")

Testing our code with previous examples

In [86]:
!chmod +x max_pairwise_product.py

!python max_pairwise_product.py < input3.txt > output3.txt
!python max_pairwise_product.py < input4.txt > output4.txt
!python max_pairwise_product.py < input5.txt > output5.txt
!python max_pairwise_product.py < input6.txt > output6.txt
!python max_pairwise_product.py < input7.txt > output7.txt
!python max_pairwise_product.py < input8.txt > output8.txt
!python max_pairwise_product.py < input9.txt > output9.txt
!python max_pairwise_product.py < input10.txt > output10.txt

/bin/bash: line 1: input6.txt: No such file or directory
/bin/bash: line 1: input7.txt: No such file or directory
/bin/bash: line 1: input8.txt: No such file or directory


In [87]:
import time

start_time = time.time()
!python max_pairwise_product.py < inputPy.txt

end_time = time.time()
execution_time = end_time - start_time
print(f"Execution time: {execution_time} segs")
print(f"Execution time: {execution_time/60} min")

Counting the steps for the python program

In [88]:
%%writefile max_pairwise_product_count.py
op = 0
xd =0
n = int(input())                        #1
op+=1
list = input().split()                  #1
op+=1
list = [int(x) for x in list]           #n
op += n

maxProduct = 0                          #1
op+=1
auxProduct = 0                          #1
op+=1
for i in range(0,n):                    #1+2n
  op+=1
  op+=2
  for j in range(i+1,n):                #((n**2)-n)/2
    op+=1
    if i!=j:                            #((n**2)-n)/2
      auxProduct = list[i]*list[j]      #((n**2)-n)/2
      op+=1
      op+=1
    op+=1
    if maxProduct < auxProduct:         #n * (n - 1) / 2
      maxProduct = auxProduct           #n * (n - 1) / 3
      op+=1

                                        # --------------For worst case
print("Result: "+str(maxProduct))
print("Number of instructions: "+str(op))

Writing max_pairwise_product_count.py


The number of instructions executed by the algooirithm with a list of size $n$ is

$$T=4 + 4n + 2(n(n-1))+(n(n-1)/3)$$

It depends on the size of the lists and the order of the list.

For the input1.txt

10

1 2 3 4 5 6 7 8 9 10

$$T(10)=255$$

In [89]:
n = 10
print(5 + 4*10+2*(n-1)*n + (n*(n-1)/3))

255.0


In [91]:
%%time
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt

### Estimate and verify the maximun n that can be solved in 1 second



In [91]:
%%time
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt

90
CPU times: user 12.2 ms, sys: 2.07 ms, total: 14.2 ms
Wall time: 813 ms


In [93]:
%%time
!python max_pairwise_product_count.py < input11.txt

In [93]:
time_per_ins = 0.001111 / 231
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction Python: {time_per_ins}")
print(f"Instructions per second Python: {instructions_per_second}")
time_per_ins = 0.00325/ 254
instructions_per_second = 1 / time_per_ins
print(f"Time per instruction C++: {time_per_ins}")
print(f"Instructions per second C++: {instructions_per_second}")

Time per instruction Python: 4.80952380952381e-06
Instructions per second Python: 207920.7920792079
Time per instruction C++: 1.279527559055118e-05
Instructions per second C++: 78153.84615384616


# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Timsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

In [94]:
%%writefile max_pairwise_product_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

// Function to calculate the maximum pairwise product using sorting

long long MaxPairwiseProduct(const std::vector<long long>& numbers) {
    int n = numbers.size();

    // Create a copy of the original vector to avoid modifying it
    std::vector<long long> sorted_numbers(numbers);

    // Sort the vector in ascending order
    std::sort(sorted_numbers.begin(), sorted_numbers.end());

    // Calculate the product of the two largest numbers
    long long max_product = sorted_numbers[n - 1] * sorted_numbers[n - 2];

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<long long> numbers(n);

    // Input the numbers
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    // Output the result
    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_sort.cpp


In [95]:
!g++ -o max_pairwise_product_sort max_pairwise_product_sort.cpp
!chmod +x max_pairwise_product_sort.cpp
!ls -l max_pairwise_product_sort.cpp max_pairwise_product_sort


-rwxr-xr-x 1 root root 40888 Dec  4 04:34 max_pairwise_product_sort
-rwxr-xr-x 1 root root   886 Dec  4 04:34 max_pairwise_product_sort.cpp


Test with inputs 1 and 2



In [96]:
!./max_pairwise_product_sort < input0.txt

0


In [97]:
!./max_pairwise_product_sort < input1.txt

90


In [98]:
!./max_pairwise_product_sort < input2.txt

90


Test and time it with input 3

In [99]:
!time ./max_pairwise_product_sort < input3.txt

6000000

real	0m0.004s
user	0m0.002s
sys	0m0.001s


20 stress tests

In [100]:
import random
import math

# Number of stress tests
num_tests = 10

for test_number in range(11, 21):
    # Determine the size of the input given the iteration
    if test_number <= 16:
        n = random.randint(2, 1000)
    else:
        n = random.randint(2, 10000)

    # Generating random values
    numbers = [random.randint(1, 99999) for _ in range(n)]

    #Writing the input in a file
    with open(f"input{test_number}.txt", "w") as file:
        file.write(f"{n}\n")
        file.write(" ".join(map(str, numbers)))

    print(f"Test {test_number}: N={n}")


Test 11: N=238
Test 12: N=824
Test 13: N=114
Test 14: N=417
Test 15: N=387
Test 16: N=312
Test 17: N=9685
Test 18: N=6170
Test 19: N=9072
Test 20: N=7678


In [101]:
!./max_pairwise_product_sort < input1.txt
!./max_pairwise_product_sort < input2.txt
!./max_pairwise_product_sort < input3.txt
!./max_pairwise_product_sort < input4.txt
!./max_pairwise_product_sort < input5.txt
!./max_pairwise_product_sort < input6.txt
!./max_pairwise_product_sort < input7.txt
!./max_pairwise_product_sort < input8.txt
!./max_pairwise_product_sort < input9.txt
!./max_pairwise_product_sort < input10.txt
!./max_pairwise_product_sort < input11.txt
!./max_pairwise_product_sort < input12.txt
!./max_pairwise_product_sort < input13.txt
!./max_pairwise_product_sort < input14.txt
!./max_pairwise_product_sort < input15.txt
!./max_pairwise_product_sort < input16.txt
!./max_pairwise_product_sort < input18.txt
!./max_pairwise_product_sort < input19.txt
!./max_pairwise_product_sort < input20.txt

90
90
6000000
31053252078
810252
/bin/bash: line 1: input6.txt: No such file or directory
/bin/bash: line 1: input7.txt: No such file or directory
/bin/bash: line 1: input8.txt: No such file or directory
7405926
37029630
9847003500
9965429865
9616816208
9900911730
9943380282
9798323150
9997600135
9994500646
9991501804


Count executed instructions in code and verify with inputs 1,2 and 3


In [102]:
%%writefile max_pairwise_product_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

//
//conteo:
//
long long MaxPairwiseProduct(const std::vector<long long>& numbers) {
    int n = numbers.size();                                                     // 1

    std::vector<long long> sorted_numbers(numbers);                             // n

    std::sort(sorted_numbers.begin(), sorted_numbers.end());                    // n*log(n)

    long long max_product = sorted_numbers[n - 1] * sorted_numbers[n - 2];      // 1

    return max_product;                                                         // 1
}

int main() {
    int n;                                                                      // 1
    std::cin >> n;                                                              // 1

    std::vector<long long> numbers(n);                                          // 1

    for (int i = 0; i < n; ++i) {                                               // n
        std::cin >> numbers[i];                                                 // n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";                           // 1

    return 0;                                                                   // 1

                                                                                // total pairwise: n+n*log(n)+3
                                                                                // total: 3n+n*log(n)+8
}



Overwriting max_pairwise_product_sort.cpp


In [103]:
%%writefile max_pairwise_product_sort2.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

long long MaxPairwiseProduct(const std::vector<long long>& numbers, long long &execints) {

    int n = numbers.size();
    execints++;

    std::vector<long long> sorted_numbers(numbers);
    execints++;

    std::sort(sorted_numbers.begin(), sorted_numbers.end());
    execints += n * log(n);

    long long max_product = sorted_numbers[n - 1] * sorted_numbers[n - 2];
    execints++;

    return max_product;
}

int main() {

    long long execints = 0LL;
    execints++;

    int n;
    execints++;

    std::cin >> n;
    execints++;

    std::vector<long long> numbers(n);
    execints += n;

    for (int i = 0; i < n; ++i) {
        execints += 2;

        std::cin >> numbers[i];
        execints++;
    }

    int M = MaxPairwiseProduct(numbers, execints);

    std::cout << execints << "\n";
    return 0;
}



Writing max_pairwise_product_sort2.cpp


In [104]:
!g++ -o max_pairwise_product_sort2 max_pairwise_product_sort2.cpp
!chmod +x max_pairwise_product_sort2.cpp
!ls -l max_pairwise_product_sort2.cpp max_pairwise_product_sort2

-rwxr-xr-x 1 root root 40928 Dec  4 04:38 max_pairwise_product_sort2
-rwxr-xr-x 1 root root   897 Dec  4 04:37 max_pairwise_product_sort2.cpp


In [105]:
n=10
print((3 * n) + (n * math.log(n)) + 8)
print(n+n * math.log(n)+3)

61.02585092994046
36.02585092994046


In [106]:
!./max_pairwise_product_sort2 < input1.txt

69


In [107]:
!./max_pairwise_product_sort2 < input2.txt

69


In [108]:
n=6
print((3 * n) + (n * math.log(n)) + 8)
print(n+n * math.log(n)+3)

36.75055681536833
19.75055681536833


In [109]:
!./max_pairwise_product_sort2 < input3.txt

40


Calculate estimated time by instruction and number of instructions by second

In [110]:
n = 200000
numbers = " ".join(map(str, range(1, 200001)))
with open('input21.txt', 'w') as file:
    file.write(f"{n}\n{numbers}\n")

In [111]:
!time ./max_pairwise_product_sort < input21.txt

39999800000

real	0m0.134s
user	0m0.122s
sys	0m0.006s


In [112]:
!./max_pairwise_product_sort2 < input21.txt

3241220


In [113]:
n=200000
print((3 * n) + (n * math.log(n)) + 8)

3041222.529106035


In [114]:
# Estimated time by instruction
0.12/3241220

3.7023096241538675e-08

In [115]:
# Estimated number of instructions by second
3241220/0.12

27010166.666666668

### Estimate and verify the maximun n that can be solved in 1 second

$$(3n + n*log(n) + 8) \times 3.7*10^{-8} s = 1 s$$

$$ n \approx 1.56×10^6 $$

In [116]:
print(1560000*(1560000-1))

2433598440000


In [117]:
n = (1560000)
numbers = " ".join(map(str, range(1, n+1)))
with open('input22.txt', 'w') as file:
    file.write(f"{n}\n{numbers}\n")

In [118]:
!time ./max_pairwise_product_sort < input22.txt

2433598440000

real	0m1.046s
user	0m1.010s
sys	0m0.027s


## Python sort





Add in each line a coment with the number of intstrcution that are executed in that line

In [119]:
%%writefile max_pairwise_product_sort.py
def maximum_pairwise_product(nums):
    # Ordenar la lista en orden descendente
    nums.sort(reverse=True)  # nlog(n)

    # Tomar los dos primeros elementos después de ordenar
    max1 = nums[0]  # 1
    max2 = nums[1]  # 1

    # Calcular el producto máximo
    result = max1 * max2  # 1

    return result  # 1


n = int(input())  # 1
input_numbers = list(map(int, input().split()))  # n
result = maximum_pairwise_product(input_numbers)  # 1
print(result)  # 1

#total: n+n*log(n)+7

Writing max_pairwise_product_sort.py


In [120]:
!chmod +x max_pairwise_product_sort.py
!ls -l max_pairwise_product_sort.py

-rwxr-xr-x 1 root root 489 Dec  4 04:44 max_pairwise_product_sort.py


In [121]:
!python max_pairwise_product_sort.py < input0.txt

0


In [122]:
!python max_pairwise_product_sort.py < input1.txt

90


In [123]:
!python max_pairwise_product_sort.py < input2.txt

90


In [124]:
!time python max_pairwise_product_sort.py < input2.txt

90

real	0m0.044s
user	0m0.025s
sys	0m0.016s


In [125]:
!python max_pairwise_product_sort.py < input1.txt
!python max_pairwise_product_sort.py < input2.txt
!python max_pairwise_product_sort.py < input3.txt
!python max_pairwise_product_sort.py < input4.txt
!python max_pairwise_product_sort.py < input5.txt
!python max_pairwise_product_sort.py < input6.txt
!python max_pairwise_product_sort.py < input7.txt
!python max_pairwise_product_sort.py < input8.txt
!python max_pairwise_product_sort.py < input9.txt
!python max_pairwise_product_sort.py < input10.txt
!python max_pairwise_product_sort.py < input11.txt
!python max_pairwise_product_sort.py < input12.txt
!python max_pairwise_product_sort.py < input13.txt
!python max_pairwise_product_sort.py < input14.txt
!python max_pairwise_product_sort.py < input15.txt
!python max_pairwise_product_sort.py < input16.txt
!python max_pairwise_product_sort.py < input17.txt
!python max_pairwise_product_sort.py < input18.txt
!python max_pairwise_product_sort.py < input19.txt
!python max_pairwise_product_sort.py < input20.txt

90
90
6000000
31053252078
810252
/bin/bash: line 1: input6.txt: No such file or directory
/bin/bash: line 1: input7.txt: No such file or directory
/bin/bash: line 1: input8.txt: No such file or directory
7405926
37029630
9847003500
9965429865
9616816208
9900911730
9943380282
9798323150
9999500006
9997600135
9994500646
9991501804


Count executed instructions in code and verify with inputs 1,2 and 3


In [126]:
%%writefile max_pairwise_product_sort2.py
import math
execints = 0

def maximum_pairwise_product(nums):
    global execints

    nums.sort(reverse=True)
    execints += n*math.log10(n)

    max1 = nums[0]
    max2 = nums[1]
    execints += 2

    result = max1 * max2  # O(1)
    execints += 1

    execints += 1
    return result

n = int(input())
execints += 1

input_numbers = list(map(int, input().split()))
execints += n

result = maximum_pairwise_product(input_numbers)
execints += 1
print(result)
execints += 1

print(execints)

Writing max_pairwise_product_sort2.py


In [127]:
!chmod +x max_pairwise_product_sort2.py
!ls -l max_pairwise_product_sort2.py

-rwxr-xr-x 1 root root 493 Dec  4 04:47 max_pairwise_product_sort2.py


In [128]:
n=10
print(n+(n*math.log10(n))+7)

27.0


In [129]:
!python max_pairwise_product_sort2.py < input1.txt

90
27.0


In [130]:
!python max_pairwise_product_sort2.py < input2.txt

90
27.0


In [131]:
!python max_pairwise_product_sort2.py < input3.txt

6000000
17.668907502301863


Calculate estimated time by instruction and number of instructions by second

In [132]:
!time python ./max_pairwise_product_sort2.py < input21.txt

39999800000
1260212.9991327962

real	0m0.094s
user	0m0.072s
sys	0m0.021s


In [133]:
n=200000
print(n+(n*math.log10(n))+7)

1260212.9991327962


In [134]:
# Estimated time by instruction
0.182/1260212.9991327962

1.4442003068151304e-07

In [135]:
# Estimated number of instructions by second
1260212.9991327962/0.12

10501774.992773302

### Estimate and verify the maximun n that can be solved in 1 second

$$(n + n*log(n) + 7) \times 1.44*10^{-7} s = 1 s$$

$$ n \approx 5×10^5 $$

*T*(N)=(n+n∗log(n)+7) is not the actual function of the code, given that the execution time is less than 1 second with the given n, the value of n that results in an execution time of 1 second is: $$ \approx 2*10 ^6$$

In [136]:
print(2000000*(2000000-1))

3999998000000


In [137]:
n = (2000000)
numbers = " ".join(map(str, range(1, n+1)))
with open('input22.txt', 'w') as file:
    file.write(f"{n}\n{numbers}\n")

In [138]:
!time python ./max_pairwise_product_sort.py < input22.txt

3999998000000

real	0m0.632s
user	0m0.441s
sys	0m0.190s


## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [139]:
%%writefile max_pairwise_product_efficient.py
def max_pairwise_product(numbers):
    max1 = float('-inf')  # 1
    max2 = float('-inf')  # 1

    for num in numbers:  # n
        if num > max1:  # n
            max2 = max1  # n
            max1 = num  # n
        elif num > max2:  # n
            max2 = num  # n

    return max1 * max2  # 1

n = int(input())  # 1

input_numbers = list(map(int, input().split()))  # n

result = max_pairwise_product(input_numbers)  # 1

print(result)  # 1

#total: 7*n+6

Writing max_pairwise_product_efficient.py


Test with inputs 1 and 2



In [140]:
!chmod +x max_pairwise_product_efficient.py
!ls -l max_pairwise_product_efficient.py

-rwxr-xr-x 1 root root 460 Dec  4 04:50 max_pairwise_product_efficient.py


Test and time it with input 3

In [141]:
!python ./max_pairwise_product_efficient.py < input0.txt
!python ./max_pairwise_product_efficient.py < input1.txt
!python ./max_pairwise_product_efficient.py < input2.txt

0
90
90


In [142]:
!time python max_pairwise_product_efficient.py < input3.txt

6000000

real	0m0.036s
user	0m0.027s
sys	0m0.009s


20 stress tests

In [143]:
!python max_pairwise_product_efficient.py < input1.txt
!python max_pairwise_product_efficient.py < input2.txt
!python max_pairwise_product_efficient.py < input3.txt
!python max_pairwise_product_efficient.py < input4.txt
!python max_pairwise_product_efficient.py < input5.txt
!python max_pairwise_product_efficient.py < input6.txt
!python max_pairwise_product_efficient.py < input7.txt
!python max_pairwise_product_efficient.py < input8.txt
!python max_pairwise_product_efficient.py < input9.txt
!python max_pairwise_product_efficient.py < input10.txt
!python max_pairwise_product_efficient.py < input11.txt
!python max_pairwise_product_efficient.py < input12.txt
!python max_pairwise_product_efficient.py < input13.txt
!python max_pairwise_product_efficient.py < input14.txt
!python max_pairwise_product_efficient.py < input15.txt
!python max_pairwise_product_efficient.py < input16.txt
!python max_pairwise_product_efficient.py < input17.txt
!python max_pairwise_product_efficient.py < input18.txt
!python max_pairwise_product_efficient.py < input19.txt
!python max_pairwise_product_efficient.py < input20.txt

90
90
6000000
31053252078
810252
/bin/bash: line 1: input6.txt: No such file or directory
/bin/bash: line 1: input7.txt: No such file or directory
/bin/bash: line 1: input8.txt: No such file or directory
7405926
37029630
9847003500
9965429865
9616816208
9900911730
9943380282
9798323150
9999500006
9997600135
9994500646
9991501804


Count executed instructions in code and verify with inputs 1,2 and 3


In [144]:
%%writefile max_pairwise_product_sort2.py
def max_pairwise_product(numbers):
    max1 = float('-inf')  # 1
    max2 = float('-inf')  # 1
    execints = 2  # Inicialización con el número de instrucciones hasta el bucle for

    for num in numbers:  # n
        if num > max1:  # n
            max2 = max1  # n
            max1 = num  # n
            execints += 3  # 3 instrucciones: comparación, asignación de max2, asignación de max1
        elif num > max2:  # n
            max2 = num  # n
            execints += 1  # 1 instrucción: asignación de max2

    return max1 * max2, execints  # 1 instrucción: retorno del producto y del número de instrucciones

n = int(input())  # 1
execints = 1  # Inicialización con el número de instrucciones hasta la lectura de n

input_numbers = list(map(int, input().split()))  # n
execints += n  # Incremento del contador de instrucciones por la creación de la lista

result, execints_func = max_pairwise_product(input_numbers)  # 1 (la función se ejecuta)
execints += execints_func  # Incremento del contador de instrucciones por la ejecución de la función

print(result)  # 1
execints += 1  # Incremento del contador de instrucciones por la impresión del resultado

print("Número total de instrucciones ejecutadas:", execints)  # 1


Overwriting max_pairwise_product_sort2.py


Calculate estimated time by instruction and number of instructions by second

In [145]:
!chmod +x max_pairwise_product_sort2.py
!ls -l max_pairwise_product_sort2.py

-rwxr-xr-x 1 root root 1248 Dec  4 04:51 max_pairwise_product_sort2.py


In [146]:
!time ./max_pairwise_product_efficient2 < input21.txt

/bin/bash: line 1: ./max_pairwise_product_efficient2: No such file or directory

real	0m0.001s
user	0m0.000s
sys	0m0.001s


### Estimate and verify the maximun n that can be solved in 1 second

$$(6*n+5) \times 6.08*10^{-8} s = 1 s$$

$$ n \approx 2.74*10^{6} $$

In [147]:
print(2740000*(2740000-1))

7507597260000


In [148]:
n = (2740000)
numbers = " ".join(map(str, range(1, n+1)))
with open('input22.txt', 'w') as file:
    file.write(f"{n}\n{numbers}\n")

In [149]:
!time ./max_pairwise_product_efficient < input22.txt

/bin/bash: line 1: ./max_pairwise_product_efficient: No such file or directory

real	0m0.004s
user	0m0.001s
sys	0m0.000s


## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [150]:
%%writefile max_pairwise_product_superefficient.cpp
#include <iostream>

using namespace std;
int main(){
  long max, secondMax, input, n;

  cin>>n;

  cin>>max;
  cin>>input;
  if(input>max){
    secondMax=max;
    max=input;
  }
  else{
    secondMax=input;
  }
  for(int i=0; i<n-2;i++){
    cin>>input;
    if(input>secondMax){
      if(input>max){
        secondMax=max;
        max=input;
      }
      else{
        secondMax=input;
      }
    }

  }
  cout<<max*secondMax<<"\n";


 cout << "Execution Time: " << time << endl;
  return 0;
}


Writing max_pairwise_product_superefficient.cpp


In [151]:
!g++ -o max_pairwise_product_superefficient max_pairwise_product_superefficient.cpp
!chmod +x max_pairwise_product_superefficient.cpp
!ls -l max_pairwise_product_superefficient.cpp max_pairwise_product_superefficient

-rwxr-xr-x 1 root root 16792 Dec  4 04:55 max_pairwise_product_superefficient
-rwxr-xr-x 1 root root   498 Dec  4 04:55 max_pairwise_product_superefficient.cpp


Test with inputs 1 and 2




In [152]:
! ./max_pairwise_product_superefficient < input1.txt
! ./max_pairwise_product_superefficient < input2.txt

90
Execution Time: 1
90
Execution Time: 1


Test and time it with input 3

In [153]:
!time ./max_pairwise_product_superefficient < input3.txt


6000000
Execution Time: 1

real	0m0.006s
user	0m0.001s
sys	0m0.002s


In [154]:
! ./max_pairwise_product_superefficient < input1.txt
! ./max_pairwise_product_superefficient < input2.txt

90
Execution Time: 1
90
Execution Time: 1


20 stress tests

In [155]:
for i in range(20,41):
  new_file=open(f"input{i}","w")

In [156]:
%%writefile randomGenerator.cpp
#include <iostream>
#include <stdio.h>
#include <unistd.h>
using namespace std;
int main (){
  int n, number;
  srand(getpid());
  n = 2+rand()%(51-2);
  cout<<n<<"\n";
  for(int i=0;i<n;i++){
    number = rand();
    cout<<number<<" ";
  }
  return 0;
}

Writing randomGenerator.cpp


In [157]:
!g++ -o randomGenerator randomGenerator.cpp
!chmod +x randomGenerator.cpp
!ls -l randomGenerator.cpp randomGenerator

-rwxr-xr-x 1 root root 16544 Dec  4 04:57 randomGenerator
-rwxr-xr-x 1 root root   255 Dec  4 04:56 randomGenerator.cpp


In [158]:
! ./randomGenerator > input20.txt
! ./randomGenerator > input21.txt
! ./randomGenerator > input22.txt
! ./randomGenerator > input23.txt
! ./randomGenerator > input24.txt
! ./randomGenerator > input25.txt
! ./randomGenerator > input26.txt
! ./randomGenerator > input27.txt
! ./randomGenerator > input28.txt
! ./randomGenerator > input29.txt
! ./randomGenerator > input30.txt
! ./randomGenerator > input31.txt
! ./randomGenerator > input32.txt
! ./randomGenerator > input33.txt
! ./randomGenerator > input34.txt
! ./randomGenerator > input35.txt
! ./randomGenerator > input36.txt
! ./randomGenerator > input37.txt
! ./randomGenerator > input38.txt
! ./randomGenerator > input39.txt
! ./randomGenerator > input40.txt

In [159]:
! ./max_pairwise_product_superefficient < input20.txt
! ./max_pairwise_product_superefficient < input21.txt
! ./max_pairwise_product_superefficient < input22.txt
! ./max_pairwise_product_superefficient < input23.txt
! ./max_pairwise_product_superefficient < input24.txt
! ./max_pairwise_product_superefficient < input25.txt
! ./max_pairwise_product_superefficient < input26.txt
! ./max_pairwise_product_superefficient < input27.txt
! ./max_pairwise_product_superefficient < input28.txt
! ./max_pairwise_product_superefficient < input29.txt
! ./max_pairwise_product_superefficient < input30.txt
! ./max_pairwise_product_superefficient < input31.txt
! ./max_pairwise_product_superefficient < input32.txt
! ./max_pairwise_product_superefficient < input33.txt
! ./max_pairwise_product_superefficient < input34.txt
! ./max_pairwise_product_superefficient < input35.txt
! ./max_pairwise_product_superefficient < input36.txt
! ./max_pairwise_product_superefficient < input37.txt
! ./max_pairwise_product_superefficient < input38.txt
! ./max_pairwise_product_superefficient < input39.txt
! ./max_pairwise_product_superefficient < input30.txt


3522159322034836896
Execution Time: 1
3962311243050339075
Execution Time: 1
4326896495333057726
Execution Time: 1
548313498334557382
Execution Time: 1
2695103839135753181
Execution Time: 1
4112505016534080672
Execution Time: 1
4157797271102837120
Execution Time: 1
4041676008175035082
Execution Time: 1
4460166371365782816
Execution Time: 1
4566256460762479027
Execution Time: 1
4084434454186551152
Execution Time: 1
4463101775653035990
Execution Time: 1
4388507348216934619
Execution Time: 1
3921709731493694840
Execution Time: 1
3952885115488536875
Execution Time: 1
4104003456119949085
Execution Time: 1
4454287758223694490
Execution Time: 1
4160776977986855010
Execution Time: 1
4469443122309335166
Execution Time: 1
4095514538052555022
Execution Time: 1
4084434454186551152
Execution Time: 1


Count executed instructions in code and verify with inputs 1,2 and 3

In [160]:
%%writefile max_pairwise_product_superefficient2.cpp
#include <iostream>
using namespace std;
int main(){
  int execints=0;
  long max, secondMax, input, n;
  cin>>n;
  execints++;
  cin>>max;
  execints++;
  cin>>input;
  execints++;
  if(input>max){
    secondMax=max;
    execints++;
    max=input;
    execints++;
  }
  else{
    secondMax=input;
    execints++;
  }
  for(int i=0; i<n-2;i++){
    cin>>input;
    execints++;
    if(input>secondMax){
      if(input>max){
        secondMax=max;
        execints++;
        max=input;
        execints++;
      }
      else{
        secondMax=input;
        execints++;
      }
    }

  }
  cout<<max*secondMax<<"\n";
  execints++;
  cout<<execints<<"\n";
  return 0;
}

Writing max_pairwise_product_superefficient2.cpp


Calculate estimated time by instruction and number of instructions by second

In [161]:
%%writefile max_pairwise_product_superefficient2.cpp
#include <iostream>
using namespace std;
int main(){
  int execints=0;
  long max, secondMax, input, n;
  cin>>n;                      #1
  cin>>max;                    #1
  cin>>input;                  #1
  if(input>max){
    secondMax=max;
    max=input;                 #2  upper bound
  }
  else{
    secondMax=input;
  }
  for(int i=0; i<n-2;i++){
    cin>>input;                #n-2
    if(input>secondMax){
      if(input>max){
        secondMax=max;
        max=input;             #2(n-2) upper bound
      }
      else{
        secondMax=input;
      }
    }

  }
  cout<<max*secondMax<<"\n";   #1
  return 0;
}

Overwriting max_pairwise_product_superefficient2.cpp


In [163]:
!g++ -o max_pairwise_product_superefficient2 max_pairwise_product_superefficient2.cpp
!chmod +x max_pairwise_product_superefficient2.cpp
!ls -l max_pairwise_product_superefficient2.cpp max_pairwise_product_superefficient2

In [164]:
! ./max_pairwise_product_superefficient2 < input1.txt

In [165]:
! ./max_pairwise_product_superefficient2 < input2.txt

In [ ]:
! ./max_pairwise_product_superefficient2 < input3.txt

##The total number of instructions (upper bound) is:
$$ f(n)=3n $$
So the algorithm complexity is O(n)

##For input1.txt n=10, f(n)=30, and our code counts exactly 30 instructions in the code, in input2.txt for n=10, our code counts 13 instructions which is within our stimates.

Calculate estimated time by instruction and number of instructions by second

In [166]:
%%writefile estimateInput.cpp
#include <iostream>
using namespace std;
int main (){
  cout << 11000<<"\n";
  for(int i=0; i<11000;i++){
      cout <<i<<" ";
  }
}

Writing estimateInput.cpp


In [167]:
!g++ -o estimateInput estimateInput.cpp
!chmod +x estimateInput.cpp
!ls -l estimateInput.cpp estimateInput

-rwxr-xr-x 1 root root 16432 Dec  4 05:01 estimateInput
-rwxr-xr-x 1 root root   133 Dec  4 05:01 estimateInput.cpp


In [168]:
new_file=open(f"results.txt","w")
new_file=open(f"superefficientInput.txt","w")

In [169]:
! ./estimateInput > superefficientInput.txt

In [170]:
!time ./max_pairwise_product_superefficient2 < superefficientInput.txt

/bin/bash: line 1: ./max_pairwise_product_superefficient2: No such file or directory

real	0m0.001s
user	0m0.000s
sys	0m0.001s


In [171]:
33000/0.001

33000000.0

### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [172]:
%%writefile max_pairwise_product_sup.py
n = int(input())  # 1
input_numbers = map(int, input().split())  # n
max1 = float('-inf')  # 1
max2 = float('-inf')  # 1

for num in input_numbers:  # n
  if num > max1:  # n
    max2 = max1  # n
    max1 = num  # n
  elif num > max2:  # n
    max2 = num  # n
#prueba
#print(max1,max2)
result = max1 * max2  # 1

print(result)  # 1

#total: 7n+5

Writing max_pairwise_product_sup.py


Test with inputs 1 and 2




In [173]:
!chmod +x max_pairwise_product_sup.py
!ls -l max_pairwise_product_sup.py

-rwxr-xr-x 1 root root 346 Dec  4 05:04 max_pairwise_product_sup.py


In [174]:
!python ./max_pairwise_product_sup.py < input0.txt

0


In [175]:
!python ./max_pairwise_product_sup.py < input1.txt

90


In [176]:
!python ./max_pairwise_product_sup.py < input2.txt

90


Test and time it with input 3

In [177]:
!time python ./max_pairwise_product_sup.py < input3.txt

6000000

real	0m0.037s
user	0m0.023s
sys	0m0.012s


20 stress tests

In [178]:
!python ./max_pairwise_product_sup.py < input1.txt
!python ./max_pairwise_product_sup.py < input2.txt
!python ./max_pairwise_product_sup.py < input3.txt
!python ./max_pairwise_product_sup.py < input4.txt
!python ./max_pairwise_product_sup.py < input5.txt
!python ./max_pairwise_product_sup.py < input6.txt
!python ./max_pairwise_product_sup.py < input7.txt
!python ./max_pairwise_product_sup.py < input8.txt
!python ./max_pairwise_product_sup.py < input9.txt
!python ./max_pairwise_product_sup.py < input10.txt
!python ./max_pairwise_product_sup.py < input11.txt
!python ./max_pairwise_product_sup.py < input12.txt
!python ./max_pairwise_product_sup.py < input13.txt
!python ./max_pairwise_product_sup.py < input15.txt
!python ./max_pairwise_product_sup.py < input16.txt
!python ./max_pairwise_product_sup.py < input17.txt
!python ./max_pairwise_product_sup.py < input18.txt
!python ./max_pairwise_product_sup.py < input19.txt
!python ./max_pairwise_product_sup.py < input20.txt

90
90
6000000
31053252078
810252
/bin/bash: line 1: input6.txt: No such file or directory
/bin/bash: line 1: input7.txt: No such file or directory
/bin/bash: line 1: input8.txt: No such file or directory
7405926
37029630
9847003500
9965429865
9616816208
9943380282
9798323150
9999500006
9997600135
9994500646
3522159322034836896


Count executed instructions in code and verify with inputs 1,2 and 3

In [179]:
%%writefile max_pairwise_product_sup2.py
execints = 0

n = int(input())  # 1
execints += 1
input_numbers = map(int, input().split())  # n
execints += 1
max1 = float('-inf')  # 1
execints += 1
max2 = float('-inf')  # 1
execints += 1

for num in input_numbers:  # n
    execints += 1
    if num > max1:  # n
        execints += 1
        max2 = max1  # n
        execints += 1
        max1 = num  # n
        execints += 1
    elif num > max2:  # n
        execints += 1
        max2 = num  # n
        execints += 1

result = max1 * max2  # 1
execints += 1

print(result)  # 1
execints += 1

print(execints)


Writing max_pairwise_product_sup2.py


In [180]:
!chmod +x max_pairwise_product_sup2.py
!ls -l max_pairwise_product_sup2.py

-rwxr-xr-x 1 root root 566 Dec  4 05:06 max_pairwise_product_sup2.py


In [181]:
n=10
print(7*n+5)

75


In [182]:
!python ./max_pairwise_product_sup2.py < input1.txt

90
46


In [183]:
!python ./max_pairwise_product_sup2.py < input2.txt

90
21


In [184]:
!python ./max_pairwise_product_sup2.py < input3.txt

6000000
20


Calculate estimated time by instruction and number of instructions by second

In [185]:
n = 200000
numbers = " ".join(map(str, range(1, 200001)))
with open('input21.txt', 'w') as file:
    file.write(f"{n}\n{numbers}\n")

In [186]:
!time python ./max_pairwise_product_sup2.py < input21.txt

39999800000
800006

real	0m0.295s
user	0m0.249s
sys	0m0.029s


In [187]:
n=200000
print(7*n+5)

1400005


In [188]:
# Estimated time by instruction
0.190/800006

2.3749821876335927e-07

In [189]:
# Estimated number of instructions by second
800006/0.190

4210557.894736842

### Estimate and verify the maximun n that can be solved in 1 second

$$(7n+5) \times 2.37*10^{-7} s = 1 s$$

$$ n \approx 6×10^5 $$

*T*(N)=(7n+5) is not the actual function of the code, given that the execution time is greater than 1 second with the given n, the value of n that results in an execution time of 1 second is: $$ \approx 2*10 ^6$$

In [190]:
print(2000000*(2000000-1))

3999998000000


In [191]:
n = (2000000)
numbers = " ".join(map(str, range(1, n+1)))
with open('input22.txt', 'w') as file:
    file.write(f"{n}\n{numbers}\n")

In [192]:
!time python ./max_pairwise_product_sup.py < input22.txt

3999998000000

real	0m0.809s
user	0m0.678s
sys	0m0.127s


## Perform 100 randomized automated tests in C++ and Python with a porgram  compares the answers,for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".

In [193]:
import random
import math

# Number of stress tests
num_tests = 100

for test_number in range(1, 101):
    n = random.randint(2, 100000)
    # Generating random values
    numbers = [random.randint(1, 99999) for _ in range(n)]

    #Writing the input in a file
    with open(f"inputltest{test_number}.txt", "w") as file:
        file.write(f"{n}\n")
        file.write(" ".join(map(str, numbers)))

    print(f"Test {test_number}: N={n}")

Test 1: N=433
Test 2: N=20576
Test 3: N=99237
Test 4: N=8775
Test 5: N=76150
Test 6: N=53600
Test 7: N=21578
Test 8: N=26011
Test 9: N=54889
Test 10: N=6214
Test 11: N=93329
Test 12: N=30917
Test 13: N=93785
Test 14: N=17791
Test 15: N=8070
Test 16: N=65474
Test 17: N=54774
Test 18: N=48072
Test 19: N=84028
Test 20: N=63813
Test 21: N=27343
Test 22: N=4687
Test 23: N=35418
Test 24: N=77506
Test 25: N=86053
Test 26: N=50153
Test 27: N=33646
Test 28: N=26925
Test 29: N=58628
Test 30: N=64237
Test 31: N=61106
Test 32: N=56179
Test 33: N=73622
Test 34: N=33537
Test 35: N=98296
Test 36: N=57180
Test 37: N=42083
Test 38: N=80445
Test 39: N=23105
Test 40: N=56467
Test 41: N=97745
Test 42: N=13586
Test 43: N=7637
Test 44: N=45495
Test 45: N=11148
Test 46: N=11570
Test 47: N=49269
Test 48: N=12057
Test 49: N=18034
Test 50: N=84984
Test 51: N=59216
Test 52: N=71549
Test 53: N=47277
Test 54: N=64053
Test 55: N=43858
Test 56: N=77259
Test 57: N=21001
Test 58: N=45963
Test 59: N=98061
Test 60: N=46

Python Super Efficient

In [194]:
# Crear un script de Python para generar líneas de comando
script_content = """\
#!/usr/bin/env python

for i in range(1, 101):
    input_file = f"inputltest{i}.txt"
    output_file = f"outputltest{i}.txt"
    command = f"python3 max_pairwise_product_sup.py < {input_file} > {output_file}"
    print(command)
"""

# Guardar el script en un archivo
with open("generate_commands.py", "w") as script_file:
    script_file.write(script_content)


In [195]:
!python generate_commands.py > run_tests.sh
!chmod +x run_tests.sh
!./run_tests.sh

In [196]:
!cat outputltest1.txt
!cat outputltest100.txt

9942183678
9996100360


C++ Super Efficient

In [197]:
script_content = """\
#!/usr/bin/env python

cpp_file = "max_pairwise_product_superefficient.cpp"
executable = "./max_pairwise_product_superefficient"

# Compilar el programa C++
compile_command = f"g++ {cpp_file} -o {executable}"
print(compile_command)

for i in range(1, 101):
    input_file = f"inputltest{i}.txt"
    output_file = f"outputctest{i}.txt"

    run_command = f"{executable} < {input_file} > {output_file}"
    print(run_command)
"""

# Guardar el script en un archivo
with open("generate_commands_cpp.py", "w") as script_file:
    script_file.write(script_content)


In [198]:
!python generate_commands_cpp.py > run_tests_c.sh
!chmod +x run_tests_c.sh
!./run_tests_c.sh

In [199]:
!cat outputctest1.txt
!cat outputctest100.txt

9942183678
Execution Time: 1
9996100360
Execution Time: 1


Test for the sort solutions python and c++:

Python Sort

In [200]:
script_content = """\
#!/usr/bin/env python

for i in range(1, 101):
    input_file = f"inputltest{i}.txt"
    output_file = f"outputpysorttest{i}.txt"
    command = f"python3 max_pairwise_product_sort.py < {input_file} > {output_file}"
    print(command)
"""

# Guardar el script en un archivo
with open("generate_commands_sort.py", "w") as script_file:
    script_file.write(script_content)


In [201]:
!python generate_commands_sort.py > run_tests_pysort.sh
!chmod +x run_tests_pysort.sh
!./run_tests_pysort.sh

In [202]:
!cat outputpysorttest1.txt
!cat outputpysorttest100.txt

9942183678
9996100360


C++ Sort

In [203]:
script_content = """\
#!/usr/bin/env python

cpp_file = "max_pairwise_product_sort.cpp"
executable = "./max_pairwise_product_sort"

# Compilar el programa C++
compile_command = f"g++ {cpp_file} -o {executable}"
print(compile_command)

for i in range(1, 101):
    input_file = f"inputltest{i}.txt"
    output_file = f"outputcsorttest{i}.txt"

    run_command = f"{executable} < {input_file} > {output_file}"
    print(run_command)
"""

# Guardar el script en un archivo
with open("generate_commands_cpp_sort.py", "w") as script_file:
    script_file.write(script_content)

In [204]:
!python generate_commands_cpp_sort.py > run_tests_c_sort.sh
!chmod +x run_tests_c_sort.sh
!./run_tests_c_sort.sh

In [205]:
!cat outputcsorttest1.txt
!cat outputcsorttest100.txt

9942183678
9996100360


Compare C++ Sort and C++ Superefficient

In [206]:
def compare_outputs(output_file1, output_file2):
    with open(output_file1, 'r') as file1, open(output_file2, 'r') as file2:
        for line1, line2 in zip(file1, file2):
            if line1.strip() != line2.strip():
                return "Error"
    return "Pass"

# Comparar los 100 archivos de salida
for i in range(1, 101):
    output_file1 = f"outputctest{i}.txt"
    output_file2 = f"outputcsorttest{i}.txt"
    result = compare_outputs(output_file1, output_file2)
    print(f"Test {i}: {result}")


Test 1: Pass
Test 2: Pass
Test 3: Pass
Test 4: Pass
Test 5: Pass
Test 6: Pass
Test 7: Pass
Test 8: Pass
Test 9: Pass
Test 10: Pass
Test 11: Pass
Test 12: Pass
Test 13: Pass
Test 14: Pass
Test 15: Pass
Test 16: Pass
Test 17: Pass
Test 18: Pass
Test 19: Pass
Test 20: Pass
Test 21: Pass
Test 22: Pass
Test 23: Pass
Test 24: Pass
Test 25: Pass
Test 26: Pass
Test 27: Pass
Test 28: Pass
Test 29: Pass
Test 30: Pass
Test 31: Pass
Test 32: Pass
Test 33: Pass
Test 34: Pass
Test 35: Pass
Test 36: Pass
Test 37: Pass
Test 38: Pass
Test 39: Pass
Test 40: Pass
Test 41: Pass
Test 42: Pass
Test 43: Pass
Test 44: Pass
Test 45: Pass
Test 46: Pass
Test 47: Pass
Test 48: Pass
Test 49: Pass
Test 50: Pass
Test 51: Pass
Test 52: Pass
Test 53: Pass
Test 54: Pass
Test 55: Pass
Test 56: Pass
Test 57: Pass
Test 58: Pass
Test 59: Pass
Test 60: Pass
Test 61: Pass
Test 62: Pass
Test 63: Pass
Test 64: Pass
Test 65: Pass
Test 66: Pass
Test 67: Pass
Test 68: Pass
Test 69: Pass
Test 70: Pass
Test 71: Pass
Test 72: Pass
T

Compare Python Sort and Python Superefficient

In [207]:
def compare_outputs(output_file1, output_file2):
    with open(output_file1, 'r') as file1, open(output_file2, 'r') as file2:
        for line1, line2 in zip(file1, file2):
            if line1.strip() != line2.strip():
                return "Error"
    return "Pass"

# Comparar los 100 archivos de salida
for i in range(1, 101):
    output_file1 = f"outputltest{i}.txt"
    output_file2 = f"outputpysorttest{i}.txt"
    result = compare_outputs(output_file1, output_file2)
    print(f"Test {i}: {result}")


Test 1: Pass
Test 2: Pass
Test 3: Pass
Test 4: Pass
Test 5: Pass
Test 6: Pass
Test 7: Pass
Test 8: Pass
Test 9: Pass
Test 10: Pass
Test 11: Pass
Test 12: Pass
Test 13: Pass
Test 14: Pass
Test 15: Pass
Test 16: Pass
Test 17: Pass
Test 18: Pass
Test 19: Pass
Test 20: Pass
Test 21: Pass
Test 22: Pass
Test 23: Pass
Test 24: Pass
Test 25: Pass
Test 26: Pass
Test 27: Pass
Test 28: Pass
Test 29: Pass
Test 30: Pass
Test 31: Pass
Test 32: Pass
Test 33: Pass
Test 34: Pass
Test 35: Pass
Test 36: Pass
Test 37: Pass
Test 38: Pass
Test 39: Pass
Test 40: Pass
Test 41: Pass
Test 42: Pass
Test 43: Pass
Test 44: Pass
Test 45: Pass
Test 46: Pass
Test 47: Pass
Test 48: Pass
Test 49: Pass
Test 50: Pass
Test 51: Pass
Test 52: Pass
Test 53: Pass
Test 54: Pass
Test 55: Pass
Test 56: Pass
Test 57: Pass
Test 58: Pass
Test 59: Pass
Test 60: Pass
Test 61: Pass
Test 62: Pass
Test 63: Pass
Test 64: Pass
Test 65: Pass
Test 66: Pass
Test 67: Pass
Test 68: Pass
Test 69: Pass
Test 70: Pass
Test 71: Pass
Test 72: Pass
T